### ISSUE TITLE
yolov3: Subsample trained weights to reduce number of classes.

### ISSUE THREAD

**Issue by hosaka** [AlexeyAB/darknet#1764]
Hi, I would like to reduce the number of predicted classes in a pre-trained yolov3 model, subsampling the classification layers. I have done this before with SSD300 using a fantastic jupyter notebook here: https://github.com/pierluigiferrari/ssd_keras/blob/master/weight_sampling_tutorial.ipynb

I am using OpenImage weights obtained from YOLO website. They are then converted to a Keras model which runs with no problem. I would like to reduce the number of predicted classes in the existing model, for example only taking single Person class. I know the class number and the subsampling code is working. An example snippet:

```
# original classifier layers, most likely missing something here
classifier_names = ['conv2d_59',
                    'conv2d_67',
                    'conv2d_75']

classes_source = 600 # total number of classes
classes_new = [68] # this is a Person class in OpenImages Dataset
```
These are pretty much all the inputs that should be necessary but I think I might be missing other classifier layers. I won't post the full code as it is very similar to the linked notebook.

Is this possible to achieve using this approach or perhaps there's a bettwe way of doing this? I know I can re-train with the original dataset and reduce it to the classes I need, but using pre-trained weights might save me some time downloading and training.

Thanks in advance!

**Comment by hosaka** [comment]
I think I am not calculating the subsample indicies correctly:
```python
for name in classifier_names:

    # get the trained weights for this layer from the source
    kernel = weights_source_file['model_weights'][name][name]['kernel:0'].value
    bias = weights_source_file['model_weights'][name][name]['bias:0'].value

    print(f"Shape of the '{name}' weights:\n")
    print(f"src_kernel:\t{kernel.shape}")
    print(f"src_bias:\t{bias.shape}")

    # get the shape of the kernel
    height, width, in_channels, out_channels = kernel.shape

    # compute the indices of the elements to subsample
    # we want to subsample relevant classes for each of the bounding boxes
    if isinstance(classes_new, (list, tuple)):
        subsample_indices = []

        for i in range(int(out_channels / classes_source)):
            indices = np.array(classes_new) + i * classes_source
            subsample_indices.append(indices)

        subsample_indices = list(np.concatenate(subsample_indices))
```

**Comment by rxjx** [comment]
Were you able to get this to work? I need to do something similar.

**Comment by hosaka** [comment]
@rxjx I wasn't no, the goal was to have this working with OpenCV, where I couldn't load SSD300. I ended up re-training darknet myself with a single class. I'd still be interested if you come up with a solution though!

**Comment by rxjx** [comment]
I was able to look into this and I think I have a solution. Here are the steps I performed:

use the ultralytics repo at https://github.com/ultralytics/yolov3 and the openimage cfg file and weights files in Alexey's repo to convert to a torch state_dict as follows:

**from models import *
m = Darknet('../yolov3-openimages.cfg', (608,608))_
m.load_state_dict(torch.load('./converted.pt', map_location='cpu')['model'])_
print(m)**

Now, I can see from the model summary that I need to adjust layers 81, 93, and 105 so and if I need to pick out class 408 out of 601 (Weapons), then the 0-based index is 407. Assuming that the 1818 weights are laid out in groups of 606, one group for each anchor, and that each group is laid out as (x, y, w, h, object-probability, class1-probability, class2-probability,....) then I need to do the following:

**for i in range(81,106,12):
      m.module_list[i][0].bias = torch.nn.Parameter(m.module_list[i][0].bias.data[[0,1,2,3,4,412, 606,607,608,609,610, 1018,1212,1213,1214,1215,1216,1624]])_
      m.module_list[i][0].weight = torch.nn.Parameter(m.module_list[i][0].weight.data[[0,1,2,3,4,412,606,607,608,609,610,1018,1212,1213,1214,1215,1216,1624]])_
      m.module_list[i][0].out_channels=18_
      m.module_list[i+1].nc=1**

That takes care of the weight surgery. Now we just write out the weights:
**save_weights(m,'./converted.weights', cutoff=-1)**

We can now use the converted.weights file with our own cfg for 1 class. I confirmed that I got a better than random result on my test set with this and that using 2 other offsets for the clas probability gave me worse results, so I suspect this might be working although I could still be wrong about the layout. 

Unfortunately, for my problem this didn't give me great results (~40% mAP) on my test set. Starting with darknet53.conv.74 got me up to 90% while using the openimage weights and replacing and retraining the last layer got me 85%. I am testing to see if training from these weights might give me a better result. Maybe @AlexeyAB can comment?
